Feature/get actual data from AI Assets

jsmatias opened the pull request:
The class `ResourceAIAssetRouter` was created inheriting `ResourceRouter`. 
The `create` method in the new class adds two new routes to the APIRouter:
    1. f"{url_prefix}/{self.resource_name_plural}/{version}/{{identifier}}/data"
    2. {url_prefix}/{self.resource_name_plural}/{version}/{{identifier}}/data/{{distribution_idx}}"
   
The `get_resource_data_func` returns a function to get the actual data from an AI asset. Setting the `default` parameter as `False`, it returns the `get_resource_data` function, which accepts the index of the distribution (i.e. 0, 1, 2, etc. for cases with multiple distribution). If `default` is set to `True` it returns the `get_resource_data_default`, which always returns the first item of the distribution list (i.e. index 0). 

Finally, the AI assets: **datasets, ml_models, experiments, and case_studies** were modified to inherit `ResourceAIAssetRouter`, instead of the previous `ResourceRouter`.

jsmatias committed:
ResourceAIAssetRouter inheriting RessourceRouter with polymorphism of the  method to append endpoints for downloading the actual data

 On branch feature/get-actual-data
 Changes to be committed:
	new file:   src/routers/resource_ai_asset_router.py
	modified:   src/routers/resource_router.py
	modified:   src/routers/resource_routers/dataset_router.py

jsmatias committed:
Generic feature to get the actual data from the AI assets.

 On branch feature/get-actual-data
 Changes to be committed:
	modified:   src/routers/resource_ai_asset_router.py
	modified:   src/routers/resource_routers/case_study_router.py
	modified:   src/routers/resource_routers/experiment_router.py
	modified:   src/routers/resource_routers/ml_model_router.py
	modified:   src/routers/resource_routers/publication_router.py
	new file:   src/tests/routers/resource_routers/test_router_case_study_retrieve_data.py
	new file:   src/tests/routers/resource_routers/test_router_dataset_retrieve_data.py
	new file:   src/tests/routers/resource_routers/test_router_experiment_retrieve_data.py
	new file:   src/tests/routers/resource_routers/test_router_ml_model_retrieve_data.py
	new file:   src/tests/routers/resource_routers/test_router_publication_retrieve_data.py

josvandervelde commented on src/tests/routers/resource_routers/test_router_case_study_retrieve_data.py:
```diff
@@ -0,0 +1,446 @@
+default_path = f"{resource_name}/v1/1/data"
+
+
+def test_get_data_from_default_with_no_distribution(
```
The different test files are functionally testing the same thing: the generic code. This code duplication is unnecessary, and makes it more difficult to make changes.

I would suggest creating one unittest file with the different scenarios (using a random one of the AIAssets). Then the only thing you still want to test is whether the CaseStudy, Dataset etc. are "configured" correctly: whether they have the endpoint. You can do that in a couple of lines per asset.

josvandervelde commented on src/tests/routers/resource_routers/test_router_case_study_retrieve_data.py:
```diff
@@ -0,0 +1,446 @@
+        session.add(person)
+        session.commit()
+
+    url_for_test = "https://zenodo.org/api/records/8206848/files/LPL11_July_2023.csv/content"
```
You're using a connection to zenodo in your unittest. This is not a good practice for a unittest: our tests should still work when Zenodo is down, when you don't have internet connection. Moreover, it makes the unittest unnecessarily slow.

I'd suggest using a RequestMock for this: you can configure that, when a GET request is performed to the `url_for_test`, a certain file is returned. See, for instance, `test_huggingface_dataset_connector.py`

jsmatias replied:
My bad. I implemented mocked responses to resolve this.

josvandervelde commented on src/routers/resource_ai_asset_router.py:
```diff
@@ -0,0 +1,99 @@
+        router = self._create(engine, url_prefix)
+
+        router.add_api_route(
+            path=f"{url_prefix}/{self.resource_name_plural}/{version}/{{identifier}}/data",
```
Now I'm seeing this, I think that "/content" might be better instead of "/data". Maybe a bit more clear, what do you think?

josvandervelde commented on src/tests/routers/resource_routers/test_router_case_study_retrieve_data.py:
```diff
@@ -0,0 +1,446 @@
+    response = client.post(
+        f"/{resource_name}/v1", json=body, headers={"Authorization": "Fake token"}
+    )
+    assert response.status_code == status.HTTP_200_OK, response.json()
```
There is quite some code duplication between, for instance `test_get_data_from_path_with_one_distribution` and `test_get_data_from_default_path_with_multiple_distributions`. This can be seen as good (keeping the test-code as simple as possible) or bad (less easy to see the differences, more difficult to make changes).
I would personally define and use some functions for the duplicated code, but this is a personal preference, I'm fine with this as well!

jsmatias replied:
I simplified the tests to made them more readable and easy to change.

josvandervelde reviewed (commented):
Looks good! I like the attention to detail (e.g. the response headers). 
I added some minor comments, only the unittests can use some real attention, I think.

jsmatias committed:
Resolved comments regarding the changes to retrieve the content from distributions and improved the structure of the automated tests
 Changes to be committed:
	modified:   src/routers/resource_ai_asset_router.py
	modified:   src/routers/resource_router.py

	new file:   src/tests/resources/contents/example1.csv
	new file:   src/tests/resources/contents/example2.tsv
	new file:   src/tests/routers/ai_asset_routers/__init__.py
	new file:   src/tests/routers/ai_asset_routers/test_router_aiassets_retrieve_content.py
	deleted:    src/tests/routers/resource_routers/test_router_case_study_retrieve_data.py
	deleted:    src/tests/routers/resource_routers/test_router_dataset_retrieve_data.py
	deleted:    src/tests/routers/resource_routers/test_router_experiment_retrieve_data.py
	deleted:    src/tests/routers/resource_routers/test_router_ml_model_retrieve_data.py
	deleted:    src/tests/routers/resource_routers/test_router_publication_retrieve_data.py

jsmatias committed:
Added pytest.mak.skip to huggingface/test_dataset_uploader.py::test_happy_path_new_repository

jsmatias committed:
Added pytest.mak.skip to another test to ignore an error for now.

jsmatias commented on src/routers/resource_ai_asset_router.py:
```diff
@@ -0,0 +1,99 @@
+        ):
+            f"""Retrieve a distribution of the actual data for {self.resource_name}
+            identified by its identifier."""
+            # 1. Get resource from id
```
My bad. I added the comments for myself and forgot to remove them.

jsmatias commented on src/routers/resource_ai_asset_router.py:
```diff
@@ -0,0 +1,99 @@
+                engine=engine, identifier=identifier, schema=schema, platform=None
+            )
+            # 2. get the url field from distribution pointing to the actual data
+            distribution = metadata.distribution  # type:ignore
```
I included this ```metadata: AIAsset = self.get_resource(``` but it doesn't really solve it. In ay case, I moved the `type: ignore` up to this line as the "root" of the issue is here.

jsmatias commented on src/routers/resource_ai_asset_router.py:
```diff
@@ -0,0 +1,99 @@
+            "tags": [self.resource_name_plural],
+        }
+
+        router = self._create(engine, url_prefix)
```
No reason. This works perfectly.  
I changed it `super(). ...`

josvandervelde commented on src/routers/resource_ai_asset_router.py:
```diff
@@ -0,0 +1,102 @@
+                    "Content-Disposition": (
+                        "attachment; " f"filename={filename or url.split('/')[-1]}"
+                    ),
+                    "Content-Type": f"{encoding_format or 'unknown'}",
```
"unknown" is not a valid mimetype. 
https://stackoverflow.com/questions/1176022/unknown-file-type-mime suggests to remove the content-type in this case, that sounds like a good approach to me!

josvandervelde reviewed (commented):
Great! Just one minor detail

jsmatias committed:
Resolved comment on the absence of a encoding format. Content-Type won't be added to headers in this case.

jsmatias committed:
Resolved conflict with pytest.mark.skip uploader/huggingface/test..

josvandervelde reviewed (approved):
Great, thanks!

jsmatias committed:
Minor change
